File: CAN/src/19/CAN_Message_Scheduling.cpp
#include <stdint.h>
#include <stdbool.h>
#include <time.h>
#include <string.h>

// CAN message structure
typedef struct {
    uint32_t id;           // CAN identifier (priority)
    uint8_t data[8];       // Data payload
    uint8_t dlc;           // Data length code
    uint32_t timestamp;    // Last transmission time (ms)
} can_message_t;

// Message schedule entry
typedef struct {
    can_message_t msg;
    uint32_t period_ms;    // Transmission period (0 = event-triggered)
    uint32_t deadline_ms;  // Message deadline
    bool (*trigger_fn)(void); // Event trigger function (NULL for time-triggered)
    void (*data_fn)(uint8_t*); // Function to populate data
} schedule_entry_t;

// Scheduling modes
typedef enum {
    SCHED_TIME_TRIGGERED,
    SCHED_EVENT_TRIGGERED,
    SCHED_HYBRID
} schedule_mode_t;

// Simple CAN transmit function (platform-specific implementation needed)
extern bool can_transmit(can_message_t* msg);

// Get current time in milliseconds
uint32_t get_time_ms(void) {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return (ts.tv_sec * 1000) + (ts.tv_nsec / 1000000);
}

// Example: Time-triggered message scheduler
void time_triggered_scheduler(schedule_entry_t* schedule, size_t count) {
    uint32_t current_time = get_time_ms();
    
    for (size_t i = 0; i < count; i++) {
        schedule_entry_t* entry = &schedule[i];
        
        // Check if message period has elapsed
        if (entry->period_ms > 0) {
            uint32_t elapsed = current_time - entry->msg.timestamp;
            
            if (elapsed >= entry->period_ms) {
                // Update message data if function provided
                if (entry->data_fn) {
                    entry->data_fn(entry->msg.data);
                }
                
                // Transmit message
                if (can_transmit(&entry->msg)) {
                    entry->msg.timestamp = current_time;
                }
            }
        }
    }
}

// Example: Event-triggered message scheduler
void event_triggered_scheduler(schedule_entry_t* schedule, size_t count) {
    uint32_t current_time = get_time_ms();
    
    for (size_t i = 0; i < count; i++) {
        schedule_entry_t* entry = &schedule[i];
        
        // Check if event condition is met
        if (entry->trigger_fn && entry->trigger_fn()) {
            // Update message data
            if (entry->data_fn) {
                entry->data_fn(entry->msg.data);
            }
            
            // Check deadline hasn't passed
            uint32_t time_since_last = current_time - entry->msg.timestamp;
            if (time_since_last < entry->deadline_ms) {
                if (can_transmit(&entry->msg)) {
                    entry->msg.timestamp = current_time;
                }
            }
        }
    }
}

// Priority queue for hybrid scheduling (simple array-based)
#define MAX_QUEUE_SIZE 32

typedef struct {
    can_message_t messages[MAX_QUEUE_SIZE];
    uint8_t count;
} priority_queue_t;

void pq_init(priority_queue_t* pq) {
    pq->count = 0;
}

// Insert message maintaining priority order (lower ID = higher priority)
bool pq_insert(priority_queue_t* pq, can_message_t* msg) {
    if (pq->count >= MAX_QUEUE_SIZE) return false;
    
    // Find insertion point
    int i = pq->count - 1;
    while (i >= 0 && pq->messages[i].id > msg->id) {
        pq->messages[i + 1] = pq->messages[i];
        i--;
    }
    
    pq->messages[i + 1] = *msg;
    pq->count++;
    return true;
}

can_message_t* pq_pop(priority_queue_t* pq) {
    if (pq->count == 0) return NULL;
    
    static can_message_t msg;
    msg = pq->messages[0];
    
    // Shift remaining messages
    for (int i = 0; i < pq->count - 1; i++) {
        pq->messages[i] = pq->messages[i + 1];
    }
    pq->count--;
    
    return &msg;
}

// Hybrid scheduler with priority queue
priority_queue_t tx_queue;

void hybrid_scheduler(schedule_entry_t* schedule, size_t count) {
    uint32_t current_time = get_time_ms();
    
    // Phase 1: Check all schedule entries
    for (size_t i = 0; i < count; i++) {
        schedule_entry_t* entry = &schedule[i];
        bool should_send = false;
        
        // Time-triggered check
        if (entry->period_ms > 0) {
            uint32_t elapsed = current_time - entry->msg.timestamp;
            if (elapsed >= entry->period_ms) {
                should_send = true;
            }
        }
        
        // Event-triggered check
        if (entry->trigger_fn && entry->trigger_fn()) {
            should_send = true;
        }
        
        if (should_send) {
            // Update data
            if (entry->data_fn) {
                entry->data_fn(entry->msg.data);
            }
            
            // Add to priority queue
            pq_insert(&tx_queue, &entry->msg);
            entry->msg.timestamp = current_time;
        }
    }
    
    // Phase 2: Transmit from priority queue
    can_message_t* msg;
    while ((msg = pq_pop(&tx_queue)) != NULL) {
        can_transmit(msg);
    }
}

// Example usage and trigger functions
bool button_pressed = false;

bool button_trigger(void) {
    bool trigger = button_pressed;
    button_pressed = false; // Clear flag
    return trigger;
}

void engine_rpm_data(uint8_t* data) {
    // Simulated RPM value: 3500 RPM
    uint16_t rpm = 3500;
    data[0] = (rpm >> 8) & 0xFF;
    data[1] = rpm & 0xFF;
}

void temperature_data(uint8_t* data) {
    // Simulated temperature: 85°C
    data[0] = 85;
}

// Application example
void setup_message_schedule(void) {
    static schedule_entry_t schedule[] = {
        // High-priority engine RPM (time-triggered, 10ms)
        {
            .msg = {.id = 0x100, .dlc = 2, .timestamp = 0},
            .period_ms = 10,
            .deadline_ms = 15,
            .trigger_fn = NULL,
            .data_fn = engine_rpm_data
        },
        // Medium-priority temperature (time-triggered, 100ms)
        {
            .msg = {.id = 0x200, .dlc = 1, .timestamp = 0},
            .period_ms = 100,
            .deadline_ms = 150,
            .trigger_fn = NULL,
            .data_fn = temperature_data
        },
        // Low-priority button event (event-triggered)
        {
            .msg = {.id = 0x400, .dlc = 1, .timestamp = 0},
            .period_ms = 0,
            .deadline_ms = 50,
            .trigger_fn = button_trigger,
            .data_fn = NULL
        }
    };
    
    pq_init(&tx_queue);
    
    // Main scheduling loop
    while (1) {
        hybrid_scheduler(schedule, 3);
        // Small delay to prevent busy-waiting
        usleep(1000); // 1ms
    }
}

File: CAN/src/19/CAN_Message_Scheduling.rs
use std::time::{Duration, Instant};
use std::collections::BinaryHeap;
use std::cmp::Ordering;

/// CAN message structure
#[derive(Debug, Clone)]
pub struct CanMessage {
    pub id: u32,
    pub data: [u8; 8],
    pub dlc: u8,
    pub timestamp: Instant,
}

impl CanMessage {
    pub fn new(id: u32) -> Self {
        Self {
            id,
            data: [0; 8],
            dlc: 0,
            timestamp: Instant::now(),
        }
    }
}

/// Priority wrapper for message queue (lower ID = higher priority)
#[derive(Debug, Clone)]
struct PriorityMessage(CanMessage);

impl PartialEq for PriorityMessage {
    fn eq(&self, other: &Self) -> bool {
        self.0.id == other.0.id
    }
}

impl Eq for PriorityMessage {}

impl PartialOrd for PriorityMessage {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for PriorityMessage {
    fn cmp(&self, other: &Self) -> Ordering {
        // Reverse ordering: lower ID = higher priority
        other.0.id.cmp(&self.0.id)
    }
}

/// Message schedule entry
pub struct ScheduleEntry {
    message: CanMessage,
    period: Option<Duration>,
    deadline: Duration,
    trigger: Option<Box<dyn Fn() -> bool + Send>>,
    data_provider: Option<Box<dyn Fn(&mut [u8; 8]) + Send>>,
}

impl ScheduleEntry {
    /// Create a time-triggered message
    pub fn time_triggered(
        id: u32,
        period: Duration,
        deadline: Duration,
        data_provider: impl Fn(&mut [u8; 8]) + Send + 'static,
    ) -> Self {
        Self {
            message: CanMessage::new(id),
            period: Some(period),
            deadline,
            trigger: None,
            data_provider: Some(Box::new(data_provider)),
        }
    }

    /// Create an event-triggered message
    pub fn event_triggered(
        id: u32,
        deadline: Duration,
        trigger: impl Fn() -> bool + Send + 'static,
        data_provider: Option<impl Fn(&mut [u8; 8]) + Send + 'static>,
    ) -> Self {
        Self {
            message: CanMessage::new(id),
            period: None,
            deadline,
            trigger: Some(Box::new(trigger)),
            data_provider: data_provider.map(|f| Box::new(f) as Box<_>),
        }
    }

    /// Create a hybrid message (both time and event triggered)
    pub fn hybrid(
        id: u32,
        period: Duration,
        deadline: Duration,
        trigger: impl Fn() -> bool + Send + 'static,
        data_provider: impl Fn(&mut [u8; 8]) + Send + 'static,
    ) -> Self {
        Self {
            message: CanMessage::new(id),
            period: Some(period),
            deadline,
            trigger: Some(Box::new(trigger)),
            data_provider: Some(Box::new(data_provider)),
        }
    }
}

/// CAN message scheduler
pub struct CanScheduler {
    entries: Vec<ScheduleEntry>,
    tx_queue: BinaryHeap<PriorityMessage>,
}

impl CanScheduler {
    pub fn new() -> Self {
        Self {
            entries: Vec::new(),
            tx_queue: BinaryHeap::new(),
        }
    }

    pub fn add_entry(&mut self, entry: ScheduleEntry) {
        self.entries.push(entry);
    }

    /// Time-triggered scheduling
    pub fn process_time_triggered(&mut self) {
        let now = Instant::now();

        for entry in &mut self.entries {
            if let Some(period) = entry.period {
                let elapsed = now.duration_since(entry.message.timestamp);

                if elapsed >= period {
                    self.prepare_message(entry, now);
                }
            }
        }
    }

    /// Event-triggered scheduling
    pub fn process_event_triggered(&mut self) {
        let now = Instant::now();

        for entry in &mut self.entries {
            if let Some(ref trigger) = entry.trigger {
                if trigger() {
                    let elapsed = now.duration_since(entry.message.timestamp);

                    // Check deadline constraint
                    if elapsed < entry.deadline {
                        self.prepare_message(entry, now);
                    }
                }
            }
        }
    }

    /// Hybrid scheduling - combines both approaches
    pub fn process_hybrid(&mut self) {
        let now = Instant::now();

        for entry in &mut self.entries {
            let mut should_send = false;

            // Check time-triggered condition
            if let Some(period) = entry.period {
                let elapsed = now.duration_since(entry.message.timestamp);
                if elapsed >= period {
                    should_send = true;
                }
            }

            // Check event-triggered condition
            if let Some(ref trigger) = entry.trigger {
                if trigger() {
                    should_send = true;
                }
            }

            if should_send {
                self.prepare_message(entry, now);
            }
        }
    }

    /// Prepare message for transmission
    fn prepare_message(&mut self, entry: &mut ScheduleEntry, now: Instant) {
        // Update data if provider exists
        if let Some(ref provider) = entry.data_provider {
            provider(&mut entry.message.data);
        }

        // Update timestamp
        entry.message.timestamp = now;

        // Add to priority queue
        self.tx_queue.push(PriorityMessage(entry.message.clone()));
    }

    /// Transmit all queued messages in priority order
    pub fn transmit_queue<F>(&mut self, mut transmit_fn: F)
    where
        F: FnMut(&CanMessage) -> Result<(), &'static str>,
    {
        while let Some(PriorityMessage(msg)) = self.tx_queue.pop() {
            match transmit_fn(&msg) {
                Ok(_) => {
                    println!("Transmitted CAN ID 0x{:03X}", msg.id);
                }
                Err(e) => {
                    eprintln!("Transmission failed: {}", e);
                    // Could re-queue or handle error
                }
            }
        }
    }

    /// Get number of queued messages
    pub fn queue_depth(&self) -> usize {
        self.tx_queue.len()
    }
}

// Example usage and data providers
fn engine_rpm_provider(data: &mut [u8; 8]) {
    let rpm: u16 = 3500; // Simulated RPM
    data[0] = (rpm >> 8) as u8;
    data[1] = (rpm & 0xFF) as u8;
}

fn temperature_provider(data: &mut [u8; 8]) {
    data[0] = 85; // 85°C
}

fn vehicle_speed_provider(data: &mut [u8; 8]) {
    let speed: u16 = 120; // 120 km/h
    data[0] = (speed >> 8) as u8;
    data[1] = (speed & 0xFF) as u8;
}

// Mock CAN transmit function
fn can_transmit(msg: &CanMessage) -> Result<(), &'static str> {
    // Platform-specific implementation would go here
    println!(
        "CAN TX: ID=0x{:03X}, DLC={}, Data={:02X?}",
        msg.id, msg.dlc, &msg.data[..msg.dlc as usize]
    );
    Ok(())
}

fn main() {
    let mut scheduler = CanScheduler::new();

    // High-priority engine RPM - time-triggered every 10ms
    scheduler.add_entry(ScheduleEntry::time_triggered(
        0x100,
        Duration::from_millis(10),
        Duration::from_millis(15),
        engine_rpm_provider,
    ));

    // Medium-priority temperature - time-triggered every 100ms
    scheduler.add_entry(ScheduleEntry::time_triggered(
        0x200,
        Duration::from_millis(100),
        Duration::from_millis(150),
        temperature_provider,
    ));

    // Low-priority speed - time-triggered every 50ms
    scheduler.add_entry(ScheduleEntry::time_triggered(
        0x300,
        Duration::from_millis(50),
        Duration::from_millis(75),
        vehicle_speed_provider,
    ));

    // Event-triggered button press
    let button_pressed = std::sync::Arc::new(std::sync::Mutex::new(false));
    let button_clone = button_pressed.clone();
    
    scheduler.add_entry(ScheduleEntry::event_triggered(
        0x400,
        Duration::from_millis(50),
        move || {
            let mut pressed = button_clone.lock().unwrap();
            let trigger = *pressed;
            *pressed = false; // Clear flag
            trigger
        },
        Some(|data: &mut [u8; 8]| {
            data[0] = 0x01; // Button press code
        }),
    ));

    // Simulation loop
    println!("Starting CAN message scheduler...\n");
    
    let start = Instant::now();
    let mut last_time = start;

    for _ in 0..100 {
        let now = Instant::now();
        
        // Simulate button press every 200ms
        if now.duration_since(last_time) >= Duration::from_millis(200) {
            *button_pressed.lock().unwrap() = true;
            last_time = now;
        }

        // Run hybrid scheduler
        scheduler.process_hybrid();

        // Transmit queued messages
        scheduler.transmit_queue(can_transmit);

        // Sleep for 1ms to simulate real-time operation
        std::thread::sleep(Duration::from_millis(1));
    }

    println!("\nScheduler demonstration complete.");
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_priority_ordering() {
        let mut queue = BinaryHeap::new();
        
        queue.push(PriorityMessage(CanMessage::new(0x300)));
        queue.push(PriorityMessage(CanMessage::new(0x100)));
        queue.push(PriorityMessage(CanMessage::new(0x200)));

        assert_eq!(queue.pop().unwrap().0.id, 0x100);
        assert_eq!(queue.pop().unwrap().0.id, 0x200);
        assert_eq!(queue.pop().unwrap().0.id, 0x300);
    }

    #[test]
    fn test_time_triggered_scheduling() {
        let mut scheduler = CanScheduler::new();
        
        scheduler.add_entry(ScheduleEntry::time_triggered(
            0x100,
            Duration::from_millis(10),
            Duration::from_millis(15),
            |_| {},
        ));

        std::thread::sleep(Duration::from_millis(11));
        scheduler.process_time_triggered();
        
        assert_eq!(scheduler.queue_depth(), 1);
    }
}